src/core/grant-authorization.ts

```diff
@@ -0,0 +1,150 @@
+    );
+
+    // Check grant scope for interface and method
+    await GrantAuthorization.verifyGrantScopeInterfaceAndMethod(incomingMessage.message, permissionsGrantMessage, permissionsGrantId);
+  }
+
+  /**
+   * Fetches PermissionsGrantMessage with CID `permissionsGrantId`, and validates that message author may use the grant
+   * for this tenant.
+   * @returns the PermissionsGrantMessage with CID `permissionsGrantId` if message exists
+   * @throws {Error} if PermissionsGrantMessage with CID `permissionsGrantId` does not exist
+   */
+  private static async fetchGrant(
+    tenant: string,
+    author: string,
+    messageStore: MessageStore,
+    permissionsGrantId: string,
+  ): Promise<PermissionsGrantMessage> {
+    const possibleGrantMessage: GenericMessage | undefined = await messageStore.get(tenant, permissionsGrantId);
+
+    const dwnInterface = possibleGrantMessage?.descriptor.interface;
+    const dwnMethod = possibleGrantMessage?.descriptor.method;
+    if (possibleGrantMessage === undefined || dwnInterface !== DwnInterfaceName.Permissions || dwnMethod !== DwnMethodName.Grant) {
+      throw new DwnError(
+        DwnErrorCode.GrantAuthorizationGrantMissing,
+        `Could not find PermissionsGrant with CID ${permissionsGrantId}`
+      );
+    }
+
+    const permissionsGrantMessage = possibleGrantMessage as PermissionsGrantMessage;
+
+    // Validate `grantedTo`
+    if (permissionsGrantMessage.descriptor.grantedTo !== author) {
+      throw new DwnError(
+        DwnErrorCode.GrantAuthorizationNotGrantedToAuthor,
+        `PermissionsGrant with CID ${permissionsGrantId} is not granted to ${author}`
+      );
+    }
+
+    // Validate `grantedFor`
+    if (permissionsGrantMessage.descriptor.grantedFor !== tenant) {
+      throw new DwnError(
+        DwnErrorCode.GrantAuthorizationNotGrantedForTenant,
+        `PermissionsGrant with CID ${permissionsGrantId} is not granted for DWN belonging to ${tenant}`
+      );
+    }
+
+    return permissionsGrantMessage;
+  }
+
+  /**
+   * Verify that the incoming message is within the allowed time frame of the grant,
+   * and the grant has not been revoked.
+   * @throws {Error} if incomingMessage has timestamp for a time in which the grant is not active.
+   */
+  public static async verifyGrantActive(
+    tenant: string,
+    incomingMessage: GenericMessage,
+    permissionsGrantMessage: PermissionsGrantMessage,
+    permissionsGrantId: string,
+    messageStore: MessageStore,
+  ): Promise<void> {
+    // Check that incomingMessage is within the grant's timeframe
+    if (incomingMessage.descriptor.messageTimestamp < permissionsGrantMessage.descriptor.messageTimestamp) {
+      // grant is not yet active
+      throw new DwnError(
+        DwnErrorCode.GrantAuthorizationGrantNotYetActive,
+        `The message has a timestamp before the associated PermissionsGrant becomes active`,
+      );
+    } else if (permissionsGrantMessage.descriptor.dateExpires !== undefined) {
```

Comment:
Consult @frankhinek and @csuwildcat if `dateExpires` should be optional.

Reply:
Discussed at office hours. None of us seem to to feel super strongly, but we leaned towards requiring `dateExpires`. It's easier to require it for now and relax the requirement later than vice versa.